File: danca_robos.java
import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;

public class danca_robos {
    private static Map<String, Integer> memo = new HashMap<>();
    private static final int n = 81;
    private static final int[] arr = {5, 9, 42, 8, 21, 79, 34, 67, 31, 15, 43, 37, 63, 73, 40, 16, 72, 25, 80, 51, 3, 12, 30, 20, 60, 62, 78, 19, 6, 14, 57, 45, 48, 58, 52, 44, 28, 11, 36, 61, 54, 2, 50, 55, 71, 77, 0, 64, 1, 18, 39, 46, 23, 65, 56, 74, 53, 22, 7, 41, 47, 59, 27, 69, 38, 35, 70, 76, 32, 4, 68, 13, 75, 10, 33, 66, 26, 24, 29, 49, 17};
    
    public static int[] troca(int[] tup, int[] pos, int n) {
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            result[i] = tup[pos[i]];
        }
        return result;
    }
    
    public static int danca(int[] arr, int n) {
        int[] chave = arr.clone();
        while (true) {
            chave = troca(chave, arr, n);
            String chaveStr = Arrays.toString(chave);
            
            if (memo.containsKey(chaveStr)) {
                memo.put(chaveStr, memo.get(chaveStr) + 1);
            } else {
                memo.put(chaveStr, 1);
            }
            
            if (memo.get(chaveStr) > 1) {
                return memo.size();
            }
        }
    }
    
    public static void main(String[] args) {
        System.out.println(danca(arr, n));
    }
}

File: danca_robos.py
n = 111
arr = [21, 86, 81, 80, 25, 64, 61, 101, 50, 94, 54, 91, 16, 7, 88, 109, 44, 32, 55, 9, 38, 41, 77, 2, 47, 53, 36, 104, 85, 11, 99, 42, 60, 14, 51, 24, 75, 17, 92, 48, 105, 67, 110, 15, 69, 71, 95, 33, 72, 6, 100, 82, 79, 102, 106, 66, 43, 22, 103, 27, 56, 59, 31, 12, 4, 96, 34, 89, 35, 87, 39, 76, 49, 57, 3, 84, 10, 108, 37, 18, 90, 58, 45, 62, 19, 26, 68, 28, 5, 0, 8, 63, 1, 65, 93, 46, 70, 107, 83, 78, 74, 97, 13, 52, 29, 73, 23, 20, 98, 30, 40]

def troca(tup: tuple[int], pos: list[int], n):
    return tuple(tup[pos[i]] for i in range(n))
        
def danca(arr, n):
    chave = arr
    memo = {tuple(arr): 1, tuple(sorted(arr)): 1}
    count = 1

    while True:
        chave = troca(chave, arr, n)
        
        if chave in memo:
            memo[chave] += 1
            if memo[chave] > 1:
                print(memo)
                return count
        count += 1

if __name__ == '__main__':
    print(danca(arr, n))
